src/hidpp-generic.c

```diff
@@ -306,6 +306,63 @@ hidpp_read_response(struct hidpp_device *dev, uint8_t *buf, size_t size)
 	return rc >= 0 ? rc : -errno;
 }
 
+int
+hidpp_test_report(struct hidpp_device *dev, uint8_t device_index, uint8_t report_type, uint report_size)
+{
+	if (report_size < 4)
+		return -EINVAL;
+
+	if (report_size > LONG_MESSAGE_LENGTH) /* this should be the longest, at least for now */
+	{
+		hidpp_log_error(dev, "We are testing a report type with size bigger than what is supported! Please report this to the libratbag developers.\n");
+		return -EINVAL;
+	}
+
+	int ret;
+	_cleanup_(freep) uint8_t *buf = zalloc(LONG_MESSAGE_LENGTH);
+
+	/* we are sending the HID++ version identification routine, please refeer to the docs for more information */
+	buf[0] = report_type;
+	buf[1] = device_index;
+	buf[2] = 0x00; /* RegisterAccessID / IRoot */
+	buf[3] = 0x10; /* ? / GetProtocolVersion() */
+
+	ret = hidpp_write_command(dev, buf, report_size);
+	if (ret)
+		return 0;
+
+	ret = hidpp_read_response(dev, buf, LONG_MESSAGE_LENGTH);
+
+	/* wait and retry if the USB timed out */
+	if (ret == -ETIMEDOUT) {
+		msleep(10);
+		ret = hidpp_read_response(dev, buf, LONG_MESSAGE_LENGTH);
+	}
+
+	/* if it didn't reply by now we assume the device does not support this report type */
+	if (ret == -ETIMEDOUT)
+		return 0;
+
+	return 1;
+}
+
+void
+hidpp_get_supported_report_types(struct hidpp_device *dev, uint8_t device_index)
+{
+	dev->supported_report_types &= (0xff << 2); /* reset the bits we are gonna check */
+
+	if (hidpp_test_report(dev, device_index, REPORT_ID_SHORT, SHORT_MESSAGE_LENGTH))
```

Comment:
I get that we might need to ensure that a device exporting a report type might not answer it, but shouldn't we also check for the `ratbag_hid_report` presence of the report ID before even trying to communicate with the device?

That might even simplify the logic of hidpp_test_report() with the timeouts and retries.

Reply:
What do you mean exactly? I should repeat the test from `hidpp_write_command`?

```
int
hidpp_write_command(struct hidpp_device *dev, uint8_t *cmd, int size)
{
	int fd = dev->hidraw_fd;
	int res;

	if (size < 1 || !cmd || fd < 0)
		return -EINVAL;
```

Reply:
nope.

`struct ratbag_hidraw` has a `reports` field which contains `struct ratbag_hid_report`. The first field in this last struct is the report ID, so I guess just by reading the static device hid definition you can see if some reports are supported or not.

Reply:
Thanks for the clarification, I will look into it.

Reply:
It does indeed solve the issue in a much better way! I didn't know we had this information available :) Thanks!